xdsl/interactive/app.py

```diff
@@ -40,17 +40,83 @@
 from xdsl.ir import MLContext
 from xdsl.parser import Parser
 from xdsl.passes import ModulePass, PipelinePass, get_pass_argument_names_and_types
+from xdsl.pattern_rewriter import PatternRewriter
 from xdsl.printer import Printer
 from xdsl.tools.command_line_tool import get_all_dialects, get_all_passes
+from xdsl.transforms import individual_rewrite
 from xdsl.transforms.mlir_opt import MLIROptPass
 from xdsl.utils.exceptions import PassPipelineParseError
 from xdsl.utils.parse_pipeline import PipelinePassSpec, parse_pipeline
 
 from ._pasteboard import pyclip_copy
 
+
+class IndividualRewrite(NamedTuple):
+    """
+    Type alias for a possible rewrite, described by an operation and pattern name.
+    """
+
+    operation: str
+    pattern: str
+
+
+class IndexedIndividualRewrite(NamedTuple):
+    """
+    Type alias for a specific rewrite pattern, additionally consisting of its operation index.
+    """
+
+    operation_index: int
+    rewrite: IndividualRewrite
+
+
 ALL_PASSES = tuple(sorted((p_name, p()) for (p_name, p) in get_all_passes().items()))
 """Contains the list of xDSL passes."""
 
+ALL_PATTERNS: tuple[IndividualRewrite, ...] = tuple(
```

Comment:
this seems not to be used

Reply:
its used in the next PR for the gui tool.

Reply:
Then can you please add it in the next PR?

Reply:
yes sure